subprojects/dependency-management/src/main/java/org/gradle/internal/component/SelectionFailureHandler.java

```diff
@@ -0,0 +1,530 @@
+    public NoMatchingVariantSelectionException noMatchingVariantsSelectionFailure(AttributesSchema schema, String displayName, ImmutableAttributes componentRequested, List<? extends ResolvedVariant> variants, AttributeMatcher matcher, AttributeDescriber attributeDescriber) {
+        String message = buildNoMatchingVariantsFailureMsg(displayName, componentRequested, variants, matcher, attributeDescriber);
+        NoMatchingVariantSelectionException e = new NoMatchingVariantSelectionException(message);
+
+        problemsService.createProblemBuilder()
+            .label("No matching variants found")
+            .undocumented()
+            .noLocation()
+            .type(FAILURE_TYPE)
+            .group(ProblemGroup.GENERIC_ID)
+            .severity(Severity.ERROR)
+            .withException(e)
+            .build();
+
+        return e;
+    }
+
+    public AmbiguousVariantSelectionException ambiguousVariantSelectionFailure(AttributesSchema schema, String displayName, ImmutableAttributes componentRequested, List<? extends ResolvedVariant> matches, AttributeMatcher matcher, Set<ResolvedVariant> discarded, AttributeDescriber attributeDescriber) {
+        String message = buildMultipleMatchingVariantsFailureMsg(attributeDescriber, displayName, componentRequested, matches, matcher, discarded);
+        AmbiguousVariantSelectionException e = new AmbiguousVariantSelectionException(message);
+
+        problemsService.createProblemBuilder()
+            .label("Multiple matching variants found")
+            .undocumented()
+            .noLocation()
+            .type(FAILURE_TYPE)
+            .group(ProblemGroup.GENERIC_ID)
+            .severity(Severity.ERROR)
+            .withException(e)
+            .build();
+
+        return e;
+    }
+
+    public AmbiguousTransformException ambiguousTransformationFailure(AttributesSchema schema, String displayName, ImmutableAttributes componentRequested, List<TransformedVariant> transformedVariants) {
+        String message = buildAmbiguousTransformMsg(displayName, componentRequested, transformedVariants);
+        AmbiguousTransformException e = new AmbiguousTransformException(message);
+
+        problemsService.createProblemBuilder()
+            .label("Ambiguous artifact transformation")
+            .undocumented()
+            .noLocation()
+            .type(FAILURE_TYPE)
+            .group(ProblemGroup.GENERIC_ID)
+            .severity(Severity.ERROR)
+            .withException(e)
+            .build();
+
+        return e;
+    }
+
+    public BrokenResolvedArtifactSet unknownSelectionFailure(AttributesSchema schema, VariantSelectionException t) {
+        problemsService.createProblemBuilder()
+            .label("Variant selection failed")
+            .undocumented()
+            .noLocation()
+            .type(FAILURE_TYPE)
+            .group(ProblemGroup.GENERIC_ID)
+            .severity(Severity.ERROR)
+            .withException(t)
+            .build();
```

Comment:
It doesn't do anything currently, but will in the future.  I could add tests that verify this is called, but they would require extensive and annoying mocking because of the stepwise builder pattern this type uses (or I add real recording functionality to the test double I built).

This is here as a placeholder for now on the advice of the IDE team.  It will be updated as that API is refined.  I'd rather wait until it's in its final form before adding unit tests to verify the calls to `Problems`.

Reply:
I certainly agree we shouldn't be mocking to verify this behavior

I just don't think we should be adding code that we don't test for, especially if it doesn't do anything

I think we should wait until this code is functional before we integrate it

Reply:
@gradle/bt-ide-experience - Can you comment on this?  I remember we had talked briefly about it being useful to add this here.

Reply:
Per Sterling's recommendation, these are left commented out for now.

Reply:
Sorry for the late reply. `build()` only creates a new problem instance of type `ReportableProblem`. You can then call `report()` on it to create a build operation progress event. You can also provided fixtures for testing reported problems in integration tests, check `AbstractIntegrationTest.enableProblemsApiCheck()`.

Reply:
Yep, I didn't want to emit operations here, just start to get the API wired in.

I'll look at what's in `AbstractIntegrationTest` when we get back to working on this and want to write comprehensive unit tests that verify the `Problems` calls.